### PR TITLE
 FIP to add Cancel FIO Request Functionality

### PR DESCRIPTION
**Pull request by edrotthoff**
This PR proposes a new FIP which specifies the addition of Cancel FIO Request functionality to the FIO Protocol API.

### CODE REVIEW

File: fip-3.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+---
+fip: 3
+title: Provide ability to cancel a request for funds
+status:Proposed
```
Comment (lukestokes) [inline fip-3.md:4]: Let's use a valid status here:
```
Draft - a FIP that is open for consideration.
Accepted - a FIP that is planned for immediate adoption.
Final - a FIP that has been adopted.
Deferred - a FIP that is not being considered for immediate adoption. May be reconsidered in the future.
```

File: fip-4.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
+---
+fip: 4
```
Comment (lukestokes) [inline fip-4.md:2]: Let's keep each FIPs in their own PRs.

File: fip-4.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+
+## Specification
+### Remove FIO address
+#### New end point: *remove_fio_address*
+##### Request
+|Parameter|Required|Format|Definition|
+|---|---|---|---|
+|fio_address|Yes|fio address, see FIO address validation rules.|FIO Address which will have public addresses removed.|
+|public_addresses|Yes|JSON Array. See public_addresses below. Min: 1 item Max: 5 item|The public address to be added to the FIO Address for the specified token.|
```
Comment (pawmmm) [inline fip-4.md:30]: Why limit to 5? It only reduces state. Should we allow to cancel all?

File: fip-4.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	
+			"status": "Ok",
+			"fee_collected": "0"		
+}
+```
+
+## Rationale
+Custom end points were put in place to make interaction with FIO Protocol easier for developers by hiding the complexity of EOSIO tools. Enhancing the functionality is done for the same reason. Advanced tools such as *get_table* remains unchanged.
+
+## Implementation
```
Comment (pawmmm) [inline fip-4.md:92]: Should also specify:
* Max transaction size
* New fee
* RAM bump (if any)

Also, I don't see Backward compatibility section. Are there any issues?

File: fip-4.md (hunk 2 above)
Comment (edrotthoff) [reply fip-4.md:2]: ill split this into independent PRs from distinct branches...

File: fip-3.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+This FIP implements the following:
+* Adds new API end point and contract action for cancellation of a request for funds.
+
+
+## Motivation
+Presently the FIO API does not provide any way for a user to cancel a request for funds they have made:
+* It is foreseeable that users will have a desire to cancel newly made requests for funds
+* When a user enters errant data (for amounts or memos)
+* When events in the users world conspire to render the request irrelevant.
+* When a users account has been "hacked" and they wish to purge the account of fraudulent requests for funds.
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:21]: ill reword, purge is a bad use of language here.

File: fip-3.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+{
+	"fio_request_id": "27",
+	"max_fee": 40000000000,
+	"tpid": ""
+        "Actor":"qwertfgewstd"
+}
+```
+##### Processing
+* Verify the request exists. Error if id is not present. 
+* Check that no payment is yet received for this request. If payment received then error.
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:45]: ill clarify the status(es) to look for in the case of payment not yet received.

File: fip-3.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+* Check that payee_fio_address is owned by actor.
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:46]: this brings up questions, do you want the pub key of the signer added as a parameter for the call?

File: fip-3.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+* require auth of the actor
+* verify that the fee for this does not exceed the max fee specified.
+* charge appropriate fee (this will be a bundled fee transaction, fee will be same as reject)
+* if there is not a status, create status to become cancelled.
+* if there is a status, and status is cancelled, then error
+* if there is a status, and the status, requested, rejected, then modify status to be cancelled.
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:52]: Ill restate so rejected is an error.

File: fip-3.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+```
+{
+	
+			"status": "cancelled",
+			"fee_collected": "0"		
+}
+```
+
+## Rationale
+Custom end points were put in place to make interaction with FIO Protocol easier for developers by hiding the complexity of EOSIO tools. Enhancing the functionality is done for the same reason. Advanced tools such as *get_table* remains unchanged.
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:78]: we are adding endpoints to the Api...so I thought the rationale was relevant...
ill also add language as to why we are adding cancel as a status.

File: fip-3.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+## Implementation
+* Add new API end point for cancel_fio_request
+  --modify chain_api_plugin to add new endpoint, modify chain_plugin.cpp and hpp to add new params and code. Add new status for cancelled to fio.request.obt.hpp. Add new action to fio.request.obt.cpp and fio.request.obt.abi. dev test api endpoint and push action and resolve all issues (2 days)
+
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:83]: no changes to get pending or get sent fio requests....
ill add a new endpoint for get_cancelled_fio_requests, I like that...nice catch.
ill add language to modify record_obt_data for a request that is cancelled..nice catch.
ill add language for the new fee determination...I thought of that but forgot to add it.
ill add language for max tx size check...
this does increase state so a ram bump is appropriate, we need to determine the amount.

File: fip-4.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+## Motivation
+Presently the FIO API does not provide any way for a user to remove public address mappings for a user:
+* It is foreseeable that users will have a need to remove pub address mappings
+* When a certain token is no longer supported
+* Whenever a user might desire to remove a certain mapping from state for privacy reasons.
+
+
+## Specification
+### Remove FIO address
+#### New end point: *remove_fio_address*
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-4.md:25]: its mentioned later in the text, ill add it here as well.

File: fip-4.md (hunk 3 above)
Comment (edrotthoff) [reply fip-4.md:30]: did this so the calls remain consistent and thus familiar....

File: fip-4.md (hunk 4 above)
Comment (edrotthoff) [reply fip-4.md:92]: ill add the max tx, new fee.

there is not a ram bump for this since its decreasing state.

File: fip-3.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pawmmm) [reply fip-3.md:46]: We can hash the pub key from request to actor and make sure actor matches, no?

File: fip-4.md (hunk 3 above)
Comment (pawmmm) [reply fip-4.md:30]: I think we need to either allow unlimited or have an ALL option, or perhaps both. This reduces state and should be encouraged.

File: fip-3.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+Presently the FIO API does not provide any way for a user to cancel a request for funds they have made:
+* It is foreseeable that users will have a desire to cancel newly made requests for funds
+* When a user enters errant data (for amounts or memos)
+* When events in the users world conspire to render the request irrelevant.
+* When a users account has been "hacked" and they wish to purge the account of fraudulent requests for funds.
+
+
+## Specification
+### Cancel Funds Request
+#### New end point: *cancel_funds_request*
```
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:26]: ill add this

File: fip-3.md (hunk 1 above)
Comment (edrotthoff) [reply fip-3.md]: yes

File: fip-3.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (edrotthoff) [reply fip-3.md:46]: ill take this as a yes...ill add canceller_pub_key as a parameter

[the remark this answers is not in the document]
Comment (pawmmm) [reply fip-3.md:46]: No, that was t osay it's not needed because you already have actor
